service/collector/trusted_advisor.go

```diff
@@ -1,167 +1,210 @@
-		Help:      "Histogram for the duration of Trusted Advisor get checks calls.",
-	})
-	getResourcesDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
-		Namespace: Namespace,
-		Name:      "trusted_advisor_get_resources_duration",
-		Help:      "Histogram for the duration of Trusted Advisor get resource calls.",
-	})
-)
+type TrustedAdvisor struct {
+	helper *helper
+	logger micrologger.Logger
+}
 
-func (c *Collector) collectAccountsTrustedAdvisorChecks(ch chan<- prometheus.Metric, clients []aws.Clients) {
-	var wg sync.WaitGroup
+func NewTrustedAdvisor(config TrustedAdvisorConfig) (*TrustedAdvisor, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
 
-	for _, client := range clients {
-		wg.Add(1)
-		go func(awsClients aws.Clients) {
-			defer wg.Done()
-			c.collectTrustedAdvisorChecks(ch, awsClients)
-		}(client)
+	t := &TrustedAdvisor{
+		helper: config.Helper,
+		logger: config.Logger,
 	}
 
-	wg.Wait()
+	return t, nil
 }
 
-func (c *Collector) collectTrustedAdvisorChecks(ch chan<- prometheus.Metric, awsClients aws.Clients) {
-	c.logger.Log("level", "debug", "message", "collecting metrics for trusted advisor checks")
-
-	checks, err := c.getTrustedAdvisorChecks(awsClients)
+func (t *TrustedAdvisor) Collect(ch chan<- prometheus.Metric) error {
+	awsClientsList, err := t.helper.GetAWSClients()
 	if err != nil {
-		if IsUnsupportedPlan(err) {
-			c.logger.Log("level", "info", "message", "Trusted Advisor not available in support plan, cannot fetch service usage metrics")
-			ch <- trustedAdvisorNotSupported()
-			return
-		}
+		return microerror.Mask(err)
+	}
+
+	var g errgroup.Group
 
-		c.logger.Log("level", "error", "message", "could not get Trusted Advisor checks", "stack", fmt.Sprintf("%#v", err))
-		trustedAdvisorError.Inc()
-		return
+	for _, item := range awsClientsList {
+		awsClients := item
+
+		g.Go(func() error {
+			err := t.collectForAccount(ch, awsClients)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return microerror.Mask(err)
 	}
-	ch <- trustedAdvisorSupported()
 
-	accountID, err := c.helper.AWSAccountID(awsClients)
+	return nil
+}
+
+func (t *TrustedAdvisor) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- serviceLimit
+	ch <- serviceUsage
+	ch <- trustedAdvisorSupport
+	return nil
+}
+
+func (t *TrustedAdvisor) collectForAccount(ch chan<- prometheus.Metric, awsClients aws.Clients) error {
+	accountID, err := t.helper.AWSAccountID(awsClients)
 	if err != nil {
-		c.logger.Log("level", "error", "message", "could not get aws account id", "stack", fmt.Sprintf("%#v", err))
+		return microerror.Mask(err)
+	}
+
+	checks, err := t.getTrustedAdvisorChecks(awsClients)
+	if IsUnsupportedPlan(err) {
+		ch <- trustedAdvisorNotSupported()
```

Comment:
During reviewing my own code, I noticed that this does not make any sense here. The collector is not executed at all when it is not supported, because we configured it that way. How should that be? I feel like we can just drop that as well.

Reply:
As stated above, you guys are more into BYOC and its implications. I would go with your advice.

Reply:
We can handle this in https://github.com/giantswarm/giantswarm/issues/3609, and consider this a refactoring PR.